Feature Guided Tour: Modeling exercise tour

lovaiible opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] ~~Server: I added multiple integration tests (Spring) related to the features~~
- [x] ~~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~~
- [x] ~~Server: I implemented the changes with a good performance and prevented too many database calls~~
- [x] ~~Server: I documented the Java code using JavaDoc style.~~
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] ~~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~~
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
This feature should show new Artemis users how to use the modeling editor: https://github.com/ls1intum/Artemis/issues/846

### Steps for Testing
1. Go to the [modeling exercise](https://artemistest.ase.in.tum.de/#/overview/38/exercises/1035) of the `Tutorial Course` on the Test Server
2. The tour should be triggered automatically
3. Navigate through the tour and complete the given tasks
4. Save the changes and start the tutorial again -> the UML model should be deleted

### Screenshots
![m1](https://user-images.githubusercontent.com/33764166/68530888-dbc78b80-030c-11ea-95b5-c07a1c392068.gif)
Reset UML model for tutorial:
![m2](https://user-images.githubusercontent.com/33764166/68530889-e08c3f80-030c-11ea-9383-a135b569ba93.gif)

lovaiible commented:
TODO:
- Create example task and check if the UML elements have been added/updated correctly (check names, attributes, methods, assocations, ...)

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -333,50 +380,48 @@ export class GuidedTourService {
-                        }
-                    });
+    private observeDomMutations(targetNode: HTMLElement, userInteraction: UserInteractionEvent, modelingTask?: string) {
+        this.mutationObserver = new MutationObserver(mutations => {
```
I'm still hoping for the observable refactoring here :)

lovaiible commented:
@thilo-behnke Thank you for reviewing my PR! I really appreciate that :) 

> Wow 👍
> 
> I found 1 issue:
> 
> * I canceled the tutorial, then went to the course administration. I was then able to start the modeling exercise tour in this area with `continue tour`:

Good that you noticed that, the error happened because of the last bug fix. I have fixed that issue now.

> Some comments:
> 
> * I had the feeling that the part with the relation did not work for me at first. I think I had to delete the relation and then try again to make it work. Not sure if I was wrong or if there was an issue.
> * I think this was already discussed, but I think it would be great to be able to restart the tour.
> * Maybe just my opinion, but it could be useful to get the solution for a step, e.g. if the student is not able to create the class correctly (due to a typo or some other careless mistake).

- I have tested the relation task with several users but could not reproduce that issue somehow. 
- The user can now restart the tour and the UML model will be removed, so that he can start from zero, after the exercise has been submitted that is not possible. I do not want to allow every user to delete their own participation, since only instructors are able to do that
- I think that would be too much overhead to embed a solution for the example exercise? The student should be able to correctly type in the UML element names or just copy them from the task description.
- Regarding the refactoring: I did that once with `Observable` and also once with `Promises` but the structure was not easier to understand, so it didn't improve the code for me. Or maybe I just misunderstood you and we can discuss that in person?

lovaiible commented:
@maxr96 can you have a look at this PR and review it? Thank you

@thilo-behnke I have changed the `mutationObserver` according to your feedback. Thanks again for reviewing

lovaiible commented on src/main/webapp/app/modeling-editor/modeling-editor.component.ts:
```diff
@@ -117,4 +125,60 @@ export class ModelingEditorComponent implements AfterViewInit, OnDestroy, OnChan
+     */
+    assessModelForGuidedTour(umlName: string, umlModel: UMLModel): void {
+        // Find the required UML classes
+        const personClass = umlModel.elements.find(element => element.name.trim() === personUML.name && element.type === 'Class');
```
Thanks for the feedback, I changed that ;)

lovaiible commented on src/main/webapp/app/modeling-editor/modeling-editor.component.ts:
```diff
@@ -117,4 +125,60 @@ export class ModelingEditorComponent implements AfterViewInit, OnDestroy, OnChan
+    assessModelForGuidedTour(umlName: string, umlModel: UMLModel): void {
+        // Find the required UML classes
+        const personClass = umlModel.elements.find(element => element.name.trim() === personUML.name && element.type === 'Class');
+        const studentClass = umlModel.elements.find(element => element.name.trim() === studentUML.name && element.type === 'Class');
```
I think in this case it is ok to not use `lodash` since we don't use complicated (calculation) logic in here that could be simplified by using lodash functions. Moreover I found this article here which would prefer using JavaScript instead of loadash for performance reasons: https://codeburst.io/why-you-shouldnt-use-lodash-anymore-and-use-pure-javascript-instead-c397df51a66

maxr96 replied:
Still I have to disagree here. The article you added is very misleading, it also tests the time used to import lodash, while we basically already pay this price once, since we are using it in other places. In real life usage lodash is just a tiny bit slower, but it checks for more corner cases, that could lead to exceptions. Here is a better article I found https://hackernoon.com/es6-lodash-native-javascript-nodejs-example-tutorial-map-function-functional-programming-performance-speed-69645e613049

lovaiible commented:
> Good work @lovaiible . Nice and clean implementation. I added only minor code and translation comments. Would be nice if somebody else can also check German translations. Some major points:

> * Vertical scroll bar is shown on all of the tutorial steps, which is a bit annoying.
> * I think it would be also nice if you finish the tutorial with actually submitting the exercise. And additionally you can mention the auto-saving mechanism. But maybe it is not for this PR.

Thank you for the review @maxr96 ! I have applied the feedback items you have mentioned.

1) I didn't notice the vertical scrollbar because on Mac it isn't shown. I fixed that :) 

2 ) I think it is ok not to submit. The user can decide if he/she wants to submit or only save the exercise. When the user leaves the exercise, a modal will pop up if the exercise is not saved/submitted yet.

ungaralex commented on src/main/webapp/app/modeling-editor/modeling-editor.component.ts:
```diff
@@ -117,4 +125,63 @@ export class ModelingEditorComponent implements AfterViewInit, OnDestroy, OnChan
+     */
+    assessModelForGuidedTour(umlName: string, umlModel: UMLModel): void {
+        // Find the required UML classes
+        const personClass = umlModel.elements.find(element => element.name.trim() === personUML.name && element.type === UMLElementType.Class);
```
This line could be refactored into its own method (since you are using pretty much the same code five times here)

ungaralex commented on src/main/webapp/app/modeling-editor/modeling-editor.component.ts:
```diff
@@ -117,4 +125,63 @@ export class ModelingEditorComponent implements AfterViewInit, OnDestroy, OnChan
+            const studentClassCorrect =
+                studentClass && majorAttribute && visitLectureMethod ? majorAttribute.owner === studentClass.id && visitLectureMethod.owner === studentClass.id : false;
+            this.guidedTourService.updateModelingResult(umlName, studentClassCorrect);
+        } else if (umlName === associationUML.name && studentClass && personClass) {
```
Nit: you can use a switch-case for the different UML names.

krusche commented:
@lovaiible Alex had some good comments, in particular the gif for the association. Could you please fix this as well?

lovaiible commented:
@ungaralex Thank you for the review! Actually step 3  of the guided tour describes and shows how to edit an association (as you can see in the screencast in the description). And the gif displayed there is this one:
https://github.com/ls1intum/Artemis/blob/f78996715b51464eaaf708069a79e6fd6d46a201/src/main/webapp/content/images/guided-tour-images/apollon-add-association.gif

@krusche Would it be ok to implement Alex' feedback on code refactoring in another PR and merge this one first?

lovaiible commented:
@maxr96  @ungaralex @krusche  I have applied the feedback and also updated the screencasts. Could you review and approve this PR? :) Thank you very much

krusche commented on src/main/webapp/app/guided-tour/guided-tour-step.model.ts:
```diff
@@ -25,6 +26,10 @@ export abstract class TourStep {
     /** Skips this step if the selector is not found, else the setStepAlreadyFinishedHint will be called by the guided tour service */
     skipStepIfNoSelector?: boolean;
+    /** List of tasks that have to be completed */
+    modelingTask?: GuidedTourModelingTask;
```
why does tour step need to be connected to modeling task? the documentation is also misleading, this is not a list.
Can we not handle this differently or more general?

There is also no `programmingTask`

lovaiible replied:
@krusche I have created an issue for that and will implement it in a follow-up PR: https://github.com/ls1intum/Artemis/issues/1062

lovaiible replied:
@krusche I changed that as discussed and created a `ModelingTaskTourStep` for that matter

krusche commented on src/main/webapp/app/guided-tour/guided-tour.model.ts:
```diff
@@ -24,4 +24,6 @@ export interface GuidedTour {
      */
     preventBackdropFromAdvancing?: boolean;
+    /** Defines if the UML model in the apollon editor should be resetted if the user restarts the tutorial */
+    resetUMLModel?: boolean;
```
this is not good, the guided tour should not depend on a `resetUMLModel` boolean. 
A better approach would be to offer a `restart` or `reset` callback that can be executed with specific steps. We discussed this already and I think before adding such attributes, we should immediately handle this properly!

lovaiible replied:
@krusche I have created an issue for that and will implement it in a follow-up PR: https://github.com/ls1intum/Artemis/issues/1062
